### ISSUE TITLE
Misleading api doc for Meteor.user

### ISSUE THREAD

**Issue by WilliamKelley** [meteor/meteor#11820]
### Issue

The API documentation of [`Meteor.user`](https://docs.meteor.com/api/accounts.html#Meteor-user) states it will return "the current user record, or null if no user is logged in." From that statement, I would assume it never returns `undefined` -- the user record would only be absent from minimongo if the user is logged-out, but then it would be `null`. However, I've found this reactive source to sometimes be `undefined`.

A correlation to my assumption is that `@types/meteor` types the function as such. I know it's a separate endeavor, but I'm mentioning it to build credence.
```ts
// "..." = omitted for brevity
module Meteor {
  interface User {
    _id: string;
    ...
  }

  function user(...): User | null;
}
```

While the documentation's function description isn't an explicit return signature, it's the closest thing to it since that's not provided either.

### Expected behavior

My expectation is that either the documentation reflects the reality: "the current user record, null if no user is logged in, and sometimes undefined while logging out -- I'll submit a change to the types package to reflect this reality. Or update the function to guarantee not returning undefined -- my personal preference.

### Reproduction

While I came across this issue formally while developing a new meteor react packages for accounts tracker hooks, [see here](https://github.com/meteor/react-packages/pull/348), I was also able to write tests within the `accounts-base` package tests that fail because of this issue. These are simpler and more useful than a reproduction repo, in my opinion, so I hope they suffice.

```js
// add these to FILE: packages/accounts-base/accounts_client_tests.js

// Proves assumption that next two tests rely on
Tinytest.addAsync(
  'accounts - Meteor.user is `null` after logout',
  (test, done) => {
    logoutAndCreateUser(test, done, () => {
      Meteor.loginWithPassword(username, password, () => {
        Meteor.logout(() => {
          test.isNull(Meteor.user());
          removeTestUser(done);
        });
      });
    });
  }
);

// Flaky, fails ~10% of runs
Tinytest.addAsync(
  'accounts - Meteor.user is never `undefined` in `Tracker.autorun` while logging out',
  (test, done) => {
    logoutAndCreateUser(test, done, () => {
      Meteor.loginWithPassword(username, password, () => {
        const computation = Tracker.autorun(() => {
          const user = Meteor.user();
          test.isNotUndefined(user);
          if (user === null) {
            computation.stop();
            removeTestUser(done);
          }
        });
        Meteor.logout();
      });
    });
  }
);

// Fails every run
Tinytest.addAsync(
  'accounts - Meteor.user is never `undefined` in `Tracker.onInvalidate` while logging out',
  (test, done) => {
    logoutAndCreateUser(test, done, () => {
      Meteor.loginWithPassword(username, password, () => {
        const computation = Tracker.autorun(() => {
          const user = Meteor.user();
          Tracker.onInvalidate(() => {
            test.isNotUndefined(Meteor.user());
          });
          if (user === null) {
            computation.stop();
            removeTestUser(done);
          }
        });
        Meteor.logout();
      });
    });
  }
);
```

In application, `Meteor.user` behaves like the "autorun" test, only returning `undefined` every once in a while. While the computation is invalidated twice during logout and `Meteor.user` returns `undefined` on the first invalidation and `null` on the second, the tracker flush only  _sometimes_ occurs before the second invalidation. There might be a better test for this, but I'm not an expert on Tracker internals and how to best manipulate them in this integration scenario.

### Root cause

An auxiliary detail is that whenever `Meteor.user` returns undefined, `Meteor.userId` returns the string of the formerly logged-in user and `Meteor.loggingOut` returns `false`. This makes sense because if the user id wasn't truthy, `Meteor.user` would return `null` because of [this condition](https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_common.js#L165-L168). In the source for [client-side `Accounts.logout`](https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_client.js#L111-L129), the order of ops is 

1. call `logout` Method
    - server runs [`logout` Method](https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_server.js#L657-L665)
      - set `userId` to `null` (server-side)
        - [auto-publish of user](https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_server.js#L767-L780) reacts and the user record is un-published (I'm not sure the exact timing on this, but it must be pretty sequential given I can produce an always-failing test!)
2. set `loggingOut` to false
3. call `makeClientLoggedOut`
    - _(aside: call `onLogout` hook(s) if `userId` isn't null)_
    - set `userId` to `null` (client-side)
4. callback

So it's clear that between 1 finishing and 3 finishing, a tracker autorun with `Meteor.user` can flush and it would return `undefined` if it received the user record pub-sub update.

Note the _aside_ of (3.); this tells me it's the intended behavior of `logout` to have `Meteor.userId` be not-null even though the server views the client as logged-out. Thus, `Meteor.user` should condition on the return of the `Meteor.users.findOne` not being undefined as well as `userId` being truthy.

_Edit:_ Another solution that's perhaps more appropriate is switching (2.) and (3.). Does it make sense that the system is no longer "logging out" the user, but the user id isn't null? And the the on-logout hooks haven't been called? Personally, I don't think so. This could simplify the return of `Meteor.user` to condition on a truthy `userId` and false `loggingOut` instead of examining the query result. However, this could be considered a breaking change.

**Comment by KoenLav** [comment]
I agree this should be improved, but not by always ensuring null.

Currently when the application is starting and we don't know whether the session cookie (resume token) is still valid the value is undefined.

This allows us to render a loading screen for authenticated routes when undefined, but redirect to login when null.

I didn't notice it being undefined when logging out (and would expect it to be null as well, but maybe there's a rationale behind this as well?

**Comment by WilliamKelley** [comment]
I don't understand your scenario. When does the user method return undefined? Why doesn't conditioning on loggingIn suffice? Login with the resume token is like any another login strategy I believe.

**Comment by WilliamKelley** [comment]
> Currently when the application is starting and we don't know whether the session cookie (resume token) is still valid the value is undefined.

Interesting, I didn't see that it [optimistically presents as logged-in by setting the user id](https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_client.js#L569-L572). I see the usefulness and it's another source of contradiction for `Meteor.user`, the user is "logged-in" but it will return `undefined`, not the user record.

> This allows us to render a loading screen for authenticated routes when undefined, but redirect to login when null.

By "this" I assume you mean the value of `Meteor.user`. I would say that's a misuse of `Meteor.user` as a "logged-in" flag rather than just record. Or, at least you could condition on `Meteor.loggingIn()` instead, because `loggingIn` would be set while the resume attempt is in progress. Would that work for you?

Overall I think we're looking for the same thing though, a ternary authentication value to which we can attribute `"not-authenticated" | "pending" | "authenticated"`. We're left hacking things together because the provided methods are binary (or claim to be), and `Meteor.user` is one that's actually ternary. I see the usefulness in keeping it that way, but I would suggest to move that kind of determination to another source instead.

**Comment by KoenLav** [comment]
We would use Meteor.loggingIn, but there's a bug there as well: https://github.com/meteor/meteor/issues/10432

Agreed on that we're looking for the same thing, not sure what the ideal change is here.

Taking into account backwards compatibility it may be preferable to keep undefined and document that it means "pending".

**Comment by WilliamKelley** [comment]
> We would use Meteor.loggingIn, but there's a bug there as well: #10432

I've left a reply. I don't think that's a bug, but the behavior isn't obvious. Can continue to discuss separately on that issue.

> ... it may be preferable to keep undefined and document that it means "pending".

I don't think it should document exactly that. It's just that `Meteor.user` can be `undefined` during logout. I'm pretty confident it's never undefined otherwise (unless the user document is deleted while the user is logged-in), because the login process is solid in calling `setUserId` on the client only after the user record publication is guaranteed to be ready.

But keeping the binary mission of `Meteor.user` seems more compelling to me. If a `"pending"` is communicated, it should be through a source built exactly with that in mind.

**Comment by KoenLav** [comment]
Ah that makes sense (loggingIn), not sure whether it's drsirable though.

I'm pretty sure Meteor.user is undefined when you are logged in, reload the page, from the moment Meteor is loaded untill loginWithToken returns.

**Comment by WilliamKelley** [comment]
> I'm pretty sure Meteor.user is undefined when you are logged in, reload the page, from the moment Meteor is loaded until loginWithToken returns. 

You're right. I'm coming to terms with the fact that as a query, `Meteor.users.findOne` has the type sig of possibly returning undefined. Meteor can make situational guarantees, but obviously those can have holes and the type shouldn't be arbitrarily narrowed.

So, I'll revise my suggestion to explicitly checking the return value of the query and returning `null` instead, if it's `undefined`. 

```js
user(options) {
    const userId = this.userId();
    const record = userId
      ? this.users.findOne(userId, this._addDefaultFieldSelector(options))
      : null;

    return record === undefined ? null : record;
  }
```

I understand we have differing opinions though.

**Comment by KoenLav** [comment]
I would indeed expect Meteor.user to return whatever Meteor.users.findOne(Meteor.userId()) would return and for us the "odd behaviour" of returning null when the user is not logged in has been helpful to determine a pending state.

If this was to be changed I think it would make more sense for Meteor.user() never to return null (since findOne wouldn't either).

I still think documenting the current behaviour is the way to go though, given backwards compatibility of this undoubtedly frequently used function.

**Comment by Kunal-Ladhani** [comment]
Hi, can I work on this. please assign it to me.

**Comment by StorytellerCZ** [comment]
@Kunal-Ladhani you don't need to ask for assignment, stating that you are working on this here is enough. Submitting PR is even more important.

**Comment by Kunal-Ladhani** [comment]
@StorytellerCZ I'm working on this issue.

**Comment by trusktr** [comment]
I've verified that the value can be `undefined`. In my testing:

- logged out: initial observable value: `null`
- do a login: value changes to `undefined` for a small moment
- logged in: value changes to user object.

I'd consider this a bug, because [the docs](https://docs.meteor.com/api/accounts#Meteor-user) state the value can only be an object, or `null`.

Answers from AIs confirm the `undefined` values:

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/3b10d10b-af5c-4eed-86b9-788075c380ce" />
<img width="400" alt="Image" src="https://github.com/user-attachments/assets/a2fcf754-fbc0-4bc0-b3f8-8f1e187947fe" />

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/ff8f7139-e567-4c7d-af08-53e78eedade6" />

**Comment by KoenLav** [comment]
Re-iterating this comment: https://github.com/meteor/meteor/issues/11820#issuecomment-1002292141

**Comment by trusktr** [comment]
@KoenLav I agree. However because Meteor owns this API, they can easily make it return `null`-only if desired:

```js
// inside Meteor's implementation:
return user ?? null // force null-only
```

I'd propose that a `Meteor.userLoading()` state would be more intuitive for that particular case too, because right now, the "loading" state is effectively unintentional from perspective of the wording in the docs.

For now, here's a PR that will update the docs and type defs:

- https://github.com/meteor/meteor/pull/14088

If we agree (and can confirm 100%) that `undefined` means "loading", then the docs should definitely state that.

**Comment by KoenLav** [comment]
Updating the docs is definitely a step in the right direction! 

Changing the behavior would mean a breaking change in one of the most frequently used functions in Meteor (across packages and applications alike). 

To me the current behavior seems sensible; for Meteor.user to return unknown as of yet (null), logged out (undefined) and logged in (user object).

**Comment by trusktr** [comment]
In the PR, I updated the type definition to add `| undefined`, which for some projects may cause a failure during build. I think either way we'll need to describe migration steps for the next release if we wish to make the API type def proper.
